packages/core/src/Cardano/types/Address.ts

```diff
@@ -1,23 +1,1259 @@
+ * │  root  │  attributes  │  type  │
+ * └────────┴──────────────┴────────┘
+ *   ╎        ╎              ╎
+ *   ╎        ╎              ╰╌╌ PubKey
+ *   ╎        ╎              ╰╌╌ Script
+ *   ╎        ╎              ╰╌╌ Redeem
+ *   ╎        ╰╌╌ Derivation Path
+ *   ╎        ╰╌╌ Network Tag
+ *   ╎
+ *   ╎                   ┌────────┬─────────────────┬──────────────┐
+ *   ╰╌╌╌╌ double-hash ( │  type  │  spending data  │  attributes  │ )
+ *                       └────────┴─────────────────┴──────────────┘
+ *                                  ╎
+ *                                  ╰╌╌ Verification Key
+ *                                  ╰╌╌ Redemption Key
+ */
+export type ByronAddressContent = {
+  root: Hash28ByteBase16;
+  attrs: ByronAttributes;
+  type: ByronAddressType;
+};
+
+/**
+ * Address object initialization properties.
+ */
+export type AddressProps = {
+  /**
+   * Specify the type of the address. There are currently 8 types of Shelley addresses and 1 for Byron addresses.
+   */
+  type: AddressType;
+
+  /**
+   * Network identifier (taken from the second half of the header (bits [3;0])).
+   *
+   * Note: not present in Bryon addresses.
+   */
+  networkId?: NetworkId;
+
+  /**
+   * Refers to a point of the chain containing a stake key registration certificate.
+   *
+   * A point is identified by 3 coordinates:
+   *
+   *  - An absolute slot number
+   *  - A transaction index (within that slot)
+   *  - A (delegation) certificate index (within that transaction)
+   */
+  pointer?: Pointer;
+
+  /**
+   * Indicates the ownership of the funds associated with the address. Whoever owns the payment parts owns any funds
+   * at the address. In order to spend from an address, one must provide a witness attesting that the address can
+   * be spent. In the case of a PubKeyHash, it means providing a signature of the transaction body made with the
+   * signing key corresponding to the hashed public key (as well as the public key itself for verification).
+   *
+   * For monetary scripts, it means being able to provide the source script and meet the necessary conditions
+   * to validate the script.
+   */
+  paymentPart?: Credential;
+
+  /**
+   * Indicates the owner of the stake rights associated with the address. Whoever owns the delegation parts owns
+   * the stake rights of any funds associated with the address. In most scenarios, the payment part and the delegation
+   * part are owned by the same party. Yet it is possible to construct addresses where both parts are owned and
+   * managed by separate entities.
+   */
+  delegationPart?: Credential;
+
+  /**
+   * Byron addresses content (kept for backward compatibility with Byron addresses).
+   */
+  byronAddressContent?: ByronAddressContent;
+};
 
 /**
  * mainnet or testnet address (Shelley as bech32 string, Byron as base58-encoded string)
  */
-export type Address = OpaqueString<'Address'>;
+export type AddressString = OpaqueString<'AddressString'>;
```

Comment:
How about thiis naming? To most people an address can't be anything other but a string :grin: 
```suggestion
export type PaymentAddress = OpaqueString<'PaymentAddress'>;
```

Reply:
Do we even need this as a separate type now? Maybe performance related? With multiple accepted encodings it seems to me that we're better off moving to just passing a class instance around, using `fromBech32` or `fromHex` static methods, and can then `toBech32` and `toHex` methods to convert the other way.

Reply:
Yes, we still need a separate opaque string for addresses. Using class objects everywhere would complicate serialization

Reply:
I'm not convinced without seeing how complicated it is, but if we do need a separate type, it should be specific: `Bech32PaymentAddress`

Reply:
This can also be Byron address. We use this type for external inputs as well, e.g. `initializeTx`

Reply:
> but to me it feels a bit redundant (and cost inefficient) to be converting from string/to string everything we interact with these types
We don't need to convert every time, once you create `Cardano.Address` you can pass it around without re-validating, so I think the boundary here is correct - it's as soon as it gets into the system, after that we trust that it has the expected format.

Reply:
If you know that it's correct type then there's no need to re-validate. `str as KeyType` should do the job?

Reply:
I think we're mixing up 2 topics here:
1. OpaqueString types - I think those are very useful and we should definitely keep them.
2. Constructors of OpaqueString types that do validation - this is debatable. We could make all validations explicit, e.g.
```ts
   type SomeKey = OpaqueString<'SomeKey'>;
   const SomeKey = (str: string): SomeKey => str as SomeKey // this doesn't necessarily have to do validation, we could do
   // 1. const SomeKey = (str: string, validate = false) => ...
   // 2. const SomeKey.validate = (str): boolean =>
   // 3. const SomeKey.parse = (str): SomeKey => // same as original SomeKey()
```

Reply:
> internal implementation could benefit from passing around objects though.

OK, I think we're in agreement here. Having the `Address` would be more convenient when doing internal transformations, which is something I'm working on now, so I can present a proposal to push the opaque string to the outer edges.